src/router.py

```diff
@@ -0,0 +1,159 @@
+from router_leg import RouterLeg    
+import pascy
+import socket
+import select
+from functools import lru_cache
+
+ROUTER_LEGS = []
+
+ARP_TABLE = {'1.1.1.2':'02:42:01:01:01:02',
+             '2.2.2.2':'02:42:02:02:02:02'}
+
+BROADCAST = 'FF:FF:FF:FF:FF:FF'
+
+def get_packet(leg : RouterLeg) -> pascy.Layer:
+    # Recieve packet
+    raw_socket = leg.raw_socket
+    pkt_raw = raw_socket.recvfrom(65536)[0]
+
+    # Parse ethernet layer
+    eth = pascy.l2.EthernetLayer()
+    eth.deserialize(pkt_raw)
+    pkt_raw = pkt_raw[len(eth):]
+
+    dst = pascy.fields.MacAddress.mac2str(eth.dst)
+    # Checks if the packet is for me
+    if dst != leg.mac and dst != BROADCAST:
+        return None
+
+    # Parse arp layer
+    if eth.ether_type == pascy.EthernetLayer.ARP_ETHER_TYPE:
+        arp = pascy.l2.ArpLayer()
+        arp.deserialize(pkt_raw)
+        pkt_raw = pkt_raw[len(arp):]
+        eth.connect_layer(arp)
+
+    # Parse ip layer
+    elif eth.ether_type == pascy.EthernetLayer.IPV4_ETHER_TYPE:
+        ip = pascy.l3.IpLayer()
+        ip.deserialize(pkt_raw)
+        pkt_raw = pkt_raw[len(ip):]
+        
+        # Parse icmp layer
+        if ip.protocol == pascy.IpLayer.ICMP_PROTOCOL_NUMBER:
+            icmp = pascy.l3.IcmpLayer()
+            icmp.deserialize(pkt_raw)
+            pkt_raw = pkt_raw[len(icmp):]
+            ip.connect_layer(icmp)
+
+        eth.connect_layer(ip)
+
+    if len(pkt_raw) > 0:
+        raw = pascy.RawLayer()
+        raw.load = pkt_raw
+        eth.connect_layer(raw)
+
+    return eth
+
+def response_arp(pkt, leg : RouterLeg):
+    resp_pkt = pascy.l2.EthernetLayer() / pascy.l2.ArpLayer() / pascy.RawLayer()
+    resp_pkt["Ethernet"].dst = pkt["Ethernet"].src
+    resp_pkt["Ethernet"].src = leg.mac
+
+    resp_pkt["ARP"].copy_layer(pkt["ARP"])
+    resp_pkt["ARP"].operation = pascy.ArpLayer.OP_IS_AT
+    resp_pkt["ARP"].sender_hardware_addr = leg.mac
+    resp_pkt["ARP"].sender_protocol_addr = leg.ip
+    resp_pkt["ARP"].target_hardware_addr = pkt["Ethernet"].src
+    resp_pkt["ARP"].target_protocol_addr = pkt["ARP"].sender_protocol_addr
+
+    leg.raw_socket.sendall(resp_pkt.build())
```

Comment:
Why `sendall`?

Reply:
Because sendall throws an exception if something failed and responsible to send the whole pkt (if the packet is too big to be sent in one send call)

Reply:
`send` doesn't throw such exceptions?

Reply:
https://stackoverflow.com/questions/34252273/what-is-the-difference-between-socket-send-and-socket-sendall

Reply:
Well this thread explains it differently than you did...
It says that `send` may send part of the buffer, while `sendall` throws an exception if it failed to send all buffer.

Reply:
I said " and responsible to send the whole pkt (if the packet is too big to be sent in one send call)" :)